File: static/components/app-shell.js
import { BaseElement, html, css } from './base.js';

class AppShell extends BaseElement {
  static properties = {
    brandName: { type: String, attribute: 'brand-name' },
    userName: { type: String, attribute: 'user-name' },
  };

  static styles = css`
    :host {
      display: grid;
      grid-template-columns: var(--sidebar-w) 1fr;
      height: 100vh;
      overflow: hidden;
    }

    .sidebar {
      background: var(--color-surface-alt);
      border-right: 1px solid var(--color-border);
      display: flex;
      flex-direction: column;
    }

    .brand {
      height: var(--topbar-h);
      padding: 0 var(--space-6);
      display: flex;
      align-items: center;
      font-weight: var(--fw-bold);
      font-size: var(--fs-lg);
      border-bottom: 1px solid var(--color-border);
    }

    .main {
      display: flex;
      flex-direction: column;
      background: var(--bg-gradient);
    }

    .topbar {
      height: var(--topbar-h);
      padding: 0 var(--space-6);
      background: var(--chrome-bg);
      backdrop-filter: var(--chrome-blur);
      border-bottom: 1px solid var(--chrome-border);
      display: flex;
      align-items: center;
      justify-content: space-between;
    }

    .content {
      flex: 1;
      padding: var(--space-8);
      overflow-y: auto;
    }

    @media (max-width: 768px) {
      :host { grid-template-columns: 1fr; }
      .sidebar { display: none; }
    }
  `;

  render() {
    return html`
      <div class="sidebar">
        <div class="brand">${this.brandName}</div>
        <slot name="sidebar"></slot>
      </div>
      <div class="main">
        <div class="topbar">
          <slot name="topbar"></slot>
          <div class="user-pill">${this.userName}</div>
        </div>
        <div class="content">
          <slot></slot>
        </div>
      </div>
    `;
  }
}

customElements.define('app-shell', AppShell);

File: src/web/ui/app_shell.rs
use super::Render;

#[derive(Default, Clone)]
pub struct AppShell {
    pub brand_name: String,
    pub user_name: Option<String>,
    pub sidebar_content: String,
    pub topbar_content: String,
    pub main_content: String,
}

impl AppShell {
    pub fn new(brand_name: impl Into<String>) -> Self {
        Self {
            brand_name: brand_name.into(),
            ..Default::default()
        }
    }

    pub fn user(mut self, name: impl Into<String>) -> Self {
        self.user_name = Some(name.into());
        self
    }

    pub fn sidebar(mut self, content: impl Into<String>) -> Self {
        self.sidebar_content = content.into();
        self
    }

    pub fn topbar(mut self, content: impl Into<String>) -> Self {
        self.topbar_content = content.into();
        self
    }

    pub fn main(mut self, content: impl Into<String>) -> Self {
        self.main_content = content.into();
        self
    }
}

impl Render for AppShell {
    fn render(&self) -> String {
        let mut attrs = Vec::new();
        attrs.push(format!("brand-name=\"{}\"", self.brand_name));
        if let Some(ref u) = self.user_name {
            attrs.push(format!("user-name=\"{}\"", u));
        }

        format!(
            "<app-shell {}>\n  <div slot=\"sidebar\">{}</div>\n  <div slot=\"topbar\">{}</div>\n  {}\n</app-shell>",
            attrs.join(" "),
            self.sidebar_content,
            self.topbar_content,
            self.main_content
        )
    }
}

impl std::fmt::Display for AppShell {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.render())
    }
}
